labstep/__init__.py

```diff
@@ -1,581 +1,646 @@
-    items.extend(resp['items'])
-    
-  return items
-
-
+    Returns
+    -------
+    entity
+        Returns the tagged entity.
+    '''
+    if 'experiments' in entity:
+        entityType = 'experiment-workflow'
+    elif 'parent_protocol' in entity:
+        entityType = 'protocol-collection'
+    elif 'resource_location' in entity:
+        entityType = 'resource'
+    elif 'collection' in entity:
+        entityType = 'protocol-collection'
+        entity = entity['collection']
+    else:
+        raise Exception('Entities of this type cannot be tagged')
+
+    url = "https://api.labstep.com/api/generic/"+entityType+"/"+str(entity['id'])+"/tag/"+str(tag['id'])
+    headers = {'apikey': user['api_key']}
+    r = requests.put(url, headers=headers)
+    return json.loads(r.content)
 
 def tag(user,entity,name):
-  '''Simple tagging of a Labstep entity (creates a new tag if none exists)
+    '''
+    Simple tagging of a Labstep entity (creates a new tag if none exists).
   
-  Parameters
-  ----------
-  user : obj
-    The labstep user to comment as. Must have property 'api_key'. See 'login'.
-  entity : obj
-    The labstep entity to tag. Can be Resource, Experiment or Protocol. Must have 'id'.
-  name : str
-    The name of the tag to add / create.
-
-  Returns
-  -------
-  entity
-    Returns the tagged entity.
-  '''
-
-  tags = getTags(user,name)
-
-  matchingTags = list(filter(lambda x: x['name']==name,tags))
-
-  if len(matchingTags)==0:
-    tag = createTag(user,name)
-  else: 
-    tag = matchingTags[0]
-
-  entity = addTagTo(user,entity,tag)
-
-  return entity
\ No newline at end of file
+    Parameters
+    ----------
+    user (obj)
+        The Labstep user to comment as. Must have property 'api_key'. See 'login'.
+    entity (obj)
+        The Labstep entity to tag. Can be Resource, Experiment, or Protocol. Must have 'id'.
+    name (str)
+        The name of the tag to create.
+
+    Returns
+    -------
+    entity
+        Returns the tagged entity.
+    '''
+    tags = getTags(user,name)
+    matchingTags = list(filter(lambda x: x['name']==name,tags))
+
+    if len(matchingTags)== 0:
+        tag = newTag(user,name)
+    else: 
+        tag = matchingTags[0]
+
+    entity = addTagTo(user,entity,tag)
+    return entity
+
+def uploadFile(user, filepath):
+    '''
+    Upload a file to the Labstep entity Data.
+
+    Parameters
+    ----------
+    user (obj)
+        The Labstep user to attach as. Must have property 'api_key'. See 'login'.
+    filepath (str)
+        The filepath to the file to attach.
+
+    Returns
+    -------
+    file
+        An object representing a file on Labstep.
+    ''' 
+    headers = {'apikey': user['api_key']}
+    files = {'file': open(filepath, 'rb')}
+    url = 'https://api.labstep.com/api/generic/file/upload'
+    r = requests.post(url, headers=headers, files=files)  
+    return json.loads(r.content)
+
+def captionUploadingFile(user,entity,filepath,caption):
```

Comment:
'captionUploadingFile' seems like the wrong name for this function. Let's call this one 'attachFile'

Reply:
Also this doesn't have a url so let's put it at the bottom.

Reply:
How about 'addFile'? To follow the app?

Reply:
yep good suggestion!